src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.cs

```diff
@@ -0,0 +1,499 @@
+                    converterInstatiationLogic = GetConverterInstantiationLogic(attributeData);
+                }
+            }
+
+            if (foundDesignTimeCustomConverter)
+            {
+                classType = converterInstatiationLogic != null
+                    ? ClassType.TypeWithDesignTimeProvidedCustomConverter
+                    : ClassType.TypeUnsupportedBySourceGen;
+            }
+            else if (_knownTypes.Contains(type))
+            {
+                classType = ClassType.KnownType;
+            }
+            else if (type.IsNullableValueType(out nullableUnderlyingType))
+            {
+                Debug.Assert(nullableUnderlyingType != null);
+                classType = ClassType.Nullable;
+            }
+            else if (type.IsEnum)
+            {
+                classType = ClassType.Enum;
+            }
+            else if (_ienumerableType.IsAssignableFrom(type))
+            {
+                // Only T[], List<T>, and Dictionary<Tkey, TValue> are supported.
+
+                if (type.IsArray)
+                {
+                    classType = ClassType.Enumerable;
+                    collectionType = CollectionType.Array;
+                    collectionValueType = type.GetElementType();
+                }
+                else if (!type.IsGenericType)
+                {
+                    classType = ClassType.TypeUnsupportedBySourceGen;
+                }
+                else
+                {
+                    Type genericTypeDef = type.GetGenericTypeDefinition();
+                    Type[] genericTypeArgs = type.GetGenericArguments();
+
+                    if (genericTypeDef == _listOfTType)
+                    {
+                        classType = ClassType.Enumerable;
+                        collectionType = CollectionType.List;
+                        collectionValueType = genericTypeArgs[0];
+                    }
+                    else if (genericTypeDef == _dictionaryType)
+                    {
+                        classType = ClassType.Dictionary;
+                        collectionType = CollectionType.Dictionary;
+                        collectionKeyType = genericTypeArgs[0];
+                        collectionValueType = genericTypeArgs[1];
+                    }
+                    else
+                    {
+                        classType = ClassType.TypeUnsupportedBySourceGen;
+                    }
+                }
+            }
+            else
+            {
+                classType = ClassType.Object;
+
+                if (type.GetConstructor(Type.EmptyTypes) != null && !type.IsAbstract && !type.IsInterface)
+                {
+                    // TODO: support parameterized ctors.
+                    constructionStrategy = ObjectConstructionStrategy.ParameterlessConstructor;
+                }
+
+                Dictionary<string, PropertyMetadata>? ignoredMembers = null;
```

Comment:
`ignoredMembers` doesn't appear to be initialized anywhere.

Reply:
It was a partial implementation to handle duplicate property names. I've pulled it now - need to support it post preview 4.